### DISCUSSION TITLE
Overriding service port and exposing http endpoints not working with preview 5

### DISCUSSION THREAD

**Discussion by murh-lego** [microsoft/aspire#3702] [General]
Hi, 

I have grpc and rest on single service. I'm overriding kestrel url and take specific ports.

Service configuration:
```
 "Kestrel__Endpoints__Rest__Url": "http://+:5000"
```

In preview 4 we used the following code to expose this port (say 5000) and it worked fine:
```
.WithEndpoint(scheme: "http", hostPort: 5000, isProxied: false)
```

Now in preview 5 with new breaking changes, it doesn't seem to work anymore with any setup I try.
My assumption is this should just straight up work:
```
  .WithExternalHttpEndpoints()
```

As I see in the dashboard env vars for my service is that the `http` endpoint is being mapped correctly, and also it's setting endpoints `IsExternal = true` flag

![image](https://github.com/dotnet/aspire/assets/149667205/17e08d57-4242-49c6-86bf-90d950423ec4)

However, I get nothing when accessing `http://localhost:5000` 😢 

What gives 🤔  I'd appreciate some guidance 😄

**Comment by mitchdenny** [comment]
Would you please be able to put together a small repro in a GitHub repo and share the link here so that we can take a look? That way we can look at the complete picture.

> NOTE: The IsExternal flag has no impact on local development experience.

**Comment by murh-lego** [comment]
@mitchdenny Repro for my case is here https://github.com/murh-lego/aspire-test/tree/main

**Comment by mitchdenny** [comment]
So what is going on here is that the web app is binding to 5000/5001 and Aspire is starting a reverse proxy on 5000 - so a conflict is occuring.

Aspire will want to dynamically assign a port to your app via the `ASPNETCORE_URLS` environment variable because it makes it possible to run a reverse proxy in front of your endpoints.

In this case you can opt out of all that by doing this:

```csharp
builder.AddProject<Projects.WebApplication1>("webapplication1", null)
       .WithHttpEndpoint(name: "rest", port: 5000, isProxied: false)
       .WithHttpEndpoint(name: "grpc", port: 5001, isProxied: false);
```

The `null` as the second argument suppresses the automatic injection of endpoints based on the launch profile. And then you provide your own but make use of the `isProxied: false` flag to disable the Aspire proxy that we launch for you.

If you want to retain the proxy you'll need to make your port allocations dynamic. The `WithEndpoint(...)` and `WithHttpEndpoint(...)` methods have variants to allow injecting the port that was allocated dynamically into an environment variable.
